zetaclient/chains/sui/signer/withdraw_and_call.go

```diff
@@ -0,0 +1,375 @@
+package signer
+
+import (
+	"context"
+	"encoding/base64"
+	"strconv"
+
+	"github.com/block-vision/sui-go-sdk/models"
+	"github.com/fardream/go-bcs/bcs"
+	"github.com/pattonkan/sui-go/sui"
+	"github.com/pattonkan/sui-go/sui/suiptb"
+	"github.com/pattonkan/sui-go/suiclient"
+	"github.com/pkg/errors"
+
+	zetasui "github.com/zeta-chain/node/pkg/contracts/sui"
+)
+
+// withdrawAndCallPTB builds unsigned withdraw and call PTB Sui transaction
+// it chains the following calls:
+// 1. withdraw_impl on gateway
+// 2. gas budget coin transfer to TSS
+// 3. on_call on target contract
+// The function returns a TxnMetaData object with tx bytes, the other fields are ignored
+func withdrawAndCallPTB(
+	signerAddrStr,
+	gatewayPackageIDStr,
+	gatewayModule string,
+	gatewayObjRef,
+	suiCoinObjRef,
+	withdrawCapObjRef sui.ObjectRef,
+	onCallObjectRefs []sui.ObjectRef,
+	coinTypeStr,
+	amountStr,
+	nonceStr,
+	gasBudgetStr,
+	receiver string,
+	cp zetasui.CallPayload,
+) (tx models.TxnMetaData, err error) {
+	ptb := suiptb.NewTransactionDataTransactionBuilder()
+
+	// Parse signer address
+	signerAddr, err := sui.AddressFromHex(signerAddrStr)
+	if err != nil {
+		return tx, errors.Wrapf(err, "invalid signer address %s", signerAddrStr)
+	}
+
+	// Add withdraw_impl command and get its command index
+	gasBudgetUint, err := addPTBCmdWithdrawImpl(
+		ptb,
+		gatewayPackageIDStr,
+		gatewayModule,
+		gatewayObjRef,
+		withdrawCapObjRef,
+		coinTypeStr,
+		amountStr,
+		nonceStr,
+		gasBudgetStr,
+	)
+	if err != nil {
+		return tx, err
+	}
+
+	// Create arguments to access the two results from the withdraw_impl call
+	cmdIndex := uint16(0)
+	argWithdrawnCoins := suiptb.Argument{
+		NestedResult: &suiptb.NestedResult{
+			Cmd:    cmdIndex,
+			Result: 0, // First result (main coins)
+		},
+	}
+
+	argBudgetCoins := suiptb.Argument{
+		NestedResult: &suiptb.NestedResult{
+			Cmd:    cmdIndex,
+			Result: 1, // Second result (budget coins)
+		},
+	}
+
+	// Add gas budget transfer command
+	err = addPTBCmdGasBudgetTransfer(ptb, argBudgetCoins, *signerAddr)
+	if err != nil {
+		return tx, err
+	}
+
+	// Add on_call command
+	err = addPTBCmdOnCall(
+		ptb,
+		receiver,
+		coinTypeStr,
+		argWithdrawnCoins,
+		onCallObjectRefs,
+		cp,
+	)
+	if err != nil {
+		return tx, err
+	}
+
+	// Finish building the PTB
+	pt := ptb.Finish()
+
+	// Wrap the PTB into a transaction data
+	txData := suiptb.NewTransactionData(
+		signerAddr,
+		pt,
+		[]*sui.ObjectRef{
+			&suiCoinObjRef,
+		},
+		gasBudgetUint,
+		suiclient.DefaultGasPrice,
```

Comment:
Why do we need the gas price here as well , conisdering the budget already includes the price 
gasBudget := strconv.FormatUint(gasPrice*params.CallOptions.GasLimit, 10)

Is this gas extra on top of what the user paid , when creating the outbound on zEVM ?

Reply:
I think this is a good point. In SUI network, a transaction needs to specify a gas budget and does not need to manually fill in the gas price, which is determined automatically by the network. The `NewTransactionData` somehow still requires a gas price parameter, so I'm assuming the `DefaultGasPrice` used here will be just a placeholder.